File: shared/sleep-calculations.ts
export const DEFAULT_SLEEP_GOAL_MINUTES = 480;

export interface SleepCalculationSession {
  id?: number;
  sleepDate?: string | null;
  bedtime?: string | Date | null;
  sleepTime?: string | Date | null;
  wakeTime?: string | Date | null;
  totalSleepDuration?: number | string | null;
  sleepEfficiency?: number | string | null;
  sleepScore?: number | string | null;
  quality?: string | null;
}

export interface SleepMetrics {
  totalSleepDuration?: number;
  timeInBedDuration?: number;
  sleepEfficiency?: number;
  sleepScore?: number;
}

export interface SleepStats extends SleepMetrics {
  dailyDuration?: number;
  dailyScore?: number;
  avgSleepDuration?: number;
  avgSleepScore?: number;
  avgEfficiency?: number;
  goalProgress?: number;
  totalSessions: number;
}

const MINUTES_PER_DAY = 24 * 60;

function asFiniteNumber(value: number | string | null | undefined): number | undefined {
  if (value === null || value === undefined || value === "") return undefined;
  const number = typeof value === "number" ? value : Number(value);
  return Number.isFinite(number) && number >= 0 ? number : undefined;
}

function asDate(value: string | Date | null | undefined): Date | undefined {
  if (value instanceof Date) {
    return Number.isNaN(value.getTime()) ? undefined : value;
  }
  if (!value) return undefined;
  const date = new Date(value);
  return Number.isNaN(date.getTime()) ? undefined : date;
}

function minutesBetween(start: string | Date | null | undefined, end: string | Date | null | undefined): number | undefined {
  const startDate = asDate(start);
  const endDate = asDate(end);
  if (!startDate || !endDate) return undefined;

  let difference = (endDate.getTime() - startDate.getTime()) / 60_000;
  if (difference <= 0) difference += MINUTES_PER_DAY;
  return Math.round(difference);
}

function clamp(value: number, minimum: number, maximum: number): number {
  return Math.min(Math.max(value, minimum), maximum);
}

function dayNumber(value: string | null | undefined): number | undefined {
  if (!value || !/^\d{4}-\d{2}-\d{2}$/.test(value)) return undefined;
  const [year, month, day] = value.split("-").map(Number);
  const date = new Date(Date.UTC(year, month - 1, day));
  return Number.isNaN(date.getTime()) ? undefined : date.getTime() / 86_400_000;
}

function round(value: number): number {
  return Math.round(value);
}

export function calculateSleepMetrics(
  session: SleepCalculationSession,
  targetSleepDuration = DEFAULT_SLEEP_GOAL_MINUTES,
): SleepMetrics {
  const calculatedDuration = minutesBetween(session.sleepTime, session.wakeTime);
  const storedDuration = asFiniteNumber(session.totalSleepDuration);
  const totalSleepDuration = calculatedDuration ?? (
    storedDuration !== undefined && storedDuration > 0 ? round(storedDuration) : undefined
  );

  const timeInBedDuration = minutesBetween(session.bedtime, session.wakeTime);
  const storedEfficiency = asFiniteNumber(session.sleepEfficiency);
  const sleepEfficiency = timeInBedDuration && totalSleepDuration !== undefined
    ? round(clamp((totalSleepDuration / timeInBedDuration) * 100, 0, 100) * 100) / 100
    : storedEfficiency !== undefined ? round(clamp(storedEfficiency, 0, 100) * 100) / 100 : undefined;

  const storedScore = asFiniteNumber(session.sleepScore);
  const sleepScore = storedScore !== undefined
    ? round(clamp(storedScore, 0, 100))
    : totalSleepDuration !== undefined && targetSleepDuration > 0
      ? round(clamp((totalSleepDuration / targetSleepDuration) * 100, 0, 100))
      : undefined;

  return {
    totalSleepDuration,
    timeInBedDuration,
    sleepEfficiency,
    sleepScore,
  };
}

export function calculateSleepStats(
  sessions: readonly SleepCalculationSession[],
  targetSleepDuration = DEFAULT_SLEEP_GOAL_MINUTES,
  referenceDate?: string,
): SleepStats | null {
  if (sessions.length === 0) return null;

  const anchor = referenceDate ?? sessions
    .map((session) => session.sleepDate)
    .filter((date): date is string => dayNumber(date) !== undefined)
    .sort()
    .at(-1);
  if (!anchor || dayNumber(anchor) === undefined) return null;

  const anchorDay = dayNumber(anchor)!;
  const metrics = sessions.map((session) => ({
    session,
    metrics: calculateSleepMetrics(session, targetSleepDuration),
  }));
  const recent = metrics.filter(({ session }) => {
    const date = dayNumber(session.sleepDate);
    return date !== undefined && date >= anchorDay - 6 && date <= anchorDay;
  });
  const selected = metrics.find(({ session }) => session.sleepDate === anchor)?.metrics;

  const durations = recent
    .map(({ metrics }) => metrics.totalSleepDuration)
    .filter((value): value is number => value !== undefined);
  const scores = metrics
    .map(({ metrics }) => metrics.sleepScore)
    .filter((value): value is number => value !== undefined);
  const efficiencies = metrics
    .map(({ metrics }) => metrics.sleepEfficiency)
    .filter((value): value is number => value !== undefined);

  const average = (values: number[]): number | undefined =>
    values.length > 0 ? round(values.reduce((sum, value) => sum + value, 0) / values.length) : undefined;

  const avgSleepDuration = average(durations);
  const avgSleepScore = average(scores);
  const avgEfficiency = average(efficiencies);

  return {
    dailyDuration: selected?.totalSleepDuration,
    dailyScore: selected?.sleepScore,
    avgSleepDuration,
    avgSleepScore,
    avgEfficiency,
    totalSleepDuration: avgSleepDuration,
    sleepScore: avgSleepScore,
    sleepEfficiency: avgEfficiency,
    goalProgress: avgSleepDuration === undefined || targetSleepDuration <= 0
      ? undefined
      : round(clamp((avgSleepDuration / targetSleepDuration) * 100, 0, 100)),
    totalSessions: durations.length,
  };
}

File: mobile_flutter/lib/features/sleep/sleep_calculations.dart
import 'models/sleep_models.dart';

const defaultSleepGoalMinutes = 480;

class SleepMetrics {
  const SleepMetrics({
    this.totalSleepDuration,
    this.timeInBedDuration,
    this.sleepEfficiency,
    this.sleepScore,
  });

  final int? totalSleepDuration;
  final int? timeInBedDuration;
  final double? sleepEfficiency;
  final int? sleepScore;
}

class SleepStats {
  const SleepStats({
    required this.avgSleepDuration,
    required this.avgSleepScore,
    required this.avgEfficiency,
    required this.totalSessions,
    required this.goalProgress,
    this.dailyDuration,
    this.dailyScore,
    this.weeklyTotalSleepDuration,
    this.weeklyAvgSleepDuration,
    this.weeklyGoalTargetMinutes,
    this.weeklyLoggedSessions,
  });

  final int? dailyDuration;
  final int? dailyScore;
  final int? avgSleepDuration;
  final int? avgSleepScore;
  final int? avgEfficiency;
  final int totalSessions;
  final int? goalProgress;
  final int? weeklyTotalSleepDuration;
  final int? weeklyAvgSleepDuration;
  final int? weeklyGoalTargetMinutes;
  final int? weeklyLoggedSessions;
}

SleepMetrics calculateSleepMetrics(
  SleepSessionModel session, {
  int targetSleepDuration = defaultSleepGoalMinutes,
}) {
  final calculatedDuration = _minutesBetween(session.sleepTime, session.wakeTime);
  final totalSleepDuration = calculatedDuration ??
      ((session.totalSleepDuration ?? 0) > 0
          ? session.totalSleepDuration!.round()
          : null);

  final timeInBedDuration = _minutesBetween(session.bedtime, session.wakeTime);
  final storedEfficiency = session.sleepEfficiency;
  final sleepEfficiency = timeInBedDuration != null &&
          timeInBedDuration > 0 &&
          totalSleepDuration != null
      ? (_clamp(totalSleepDuration / timeInBedDuration * 100, 0, 100) * 100)
              .round() /
          100
      : storedEfficiency == null
          ? null
          : (_clamp(storedEfficiency, 0, 100) * 100).round() / 100;

  final storedScore = session.sleepScore;
  final sleepScore = storedScore != null
      ? _round(_clamp(storedScore.toDouble(), 0, 100))
      : totalSleepDuration != null && targetSleepDuration > 0
          ? _round(
              _clamp(totalSleepDuration / targetSleepDuration * 100, 0, 100),
            )
          : null;

  return SleepMetrics(
    totalSleepDuration: totalSleepDuration,
    timeInBedDuration: timeInBedDuration,
    sleepEfficiency: sleepEfficiency,
    sleepScore: sleepScore,
  );
}

SleepStats? calculateSleepStats(
  List<SleepSessionModel> sessions, {
  int targetSleepDuration = defaultSleepGoalMinutes,
  String? referenceDate,
}) {
  if (sessions.isEmpty) return null;
  final datedSessions = sessions
      .map((session) => session.sleepDate)
      .where(_isValidDateOnly)
      .toList()
    ..sort();
  final anchor = referenceDate ?? (datedSessions.isEmpty ? null : datedSessions.last);
  final anchorDay = anchor == null ? null : _dayNumber(anchor);
  if (anchorDay == null) return null;

  final metricSessions = sessions
      .map(
        (session) => (
          session: session,
          metrics: calculateSleepMetrics(
            session,
            targetSleepDuration: targetSleepDuration,
          ),
        ),
      )
      .toList();
  final recent = metricSessions.where((item) {
    final day = _dayNumber(item.session.sleepDate);
    return day != null && day >= anchorDay - 6 && day <= anchorDay;
  }).toList();
  final weekStart = anchorDay - _weekdaySundayZero(anchorDay);
  final weekly = metricSessions.where((item) {
    final day = _dayNumber(item.session.sleepDate);
    return day != null && day >= weekStart && day <= anchorDay;
  }).toList();

  SleepMetrics? selected;
  for (final item in metricSessions) {
    if (item.session.sleepDate == anchor) {
      selected = item.metrics;
      break;
    }
  }
  final durations = recent
      .map((item) => item.metrics.totalSleepDuration)
      .whereType<int>()
      .toList();
  final scores = metricSessions
      .map((item) => item.metrics.sleepScore)
      .whereType<int>()
      .toList();
  final efficiencies = metricSessions
      .map((item) => item.metrics.sleepEfficiency)
      .whereType<double>()
      .toList();
  final weeklyDurations = weekly
      .map((item) => item.metrics.totalSleepDuration)
      .whereType<int>()
      .toList();
  final weeklyGoalTarget = targetSleepDuration > 0
      ? targetSleepDuration * (anchorDay - weekStart + 1)
      : null;

  return SleepStats(
    dailyDuration: selected?.totalSleepDuration,
    dailyScore: selected?.sleepScore,
    avgSleepDuration: _average(durations),
    avgSleepScore: _average(scores),
    avgEfficiency: _average(efficiencies),
    goalProgress: weeklyGoalTarget == null || weeklyGoalTarget == 0
        ? null
        : _round(
            _clamp(
              weeklyDurations.fold<int>(0, (sum, value) => sum + value) /
                  weeklyGoalTarget *
                  100,
              0,
              100,
            ),
          ),
    weeklyTotalSleepDuration:
        weeklyDurations.fold<int>(0, (sum, value) => sum + value),
    weeklyAvgSleepDuration: _average(weeklyDurations),
    weeklyGoalTargetMinutes: weeklyGoalTarget,
    weeklyLoggedSessions: weeklyDurations.length,
    totalSessions: durations.length,
  );
}

int? _average(List<num> values) {
  if (values.isEmpty) return null;
  final total = values.fold<num>(0, (sum, value) => sum + value);
  return _round(total / values.length);
}

int _weekdaySundayZero(int dayNumber) {
  final date = DateTime.utc(1970, 1, 1).add(Duration(days: dayNumber));
  return date.weekday % 7;
}

int? _dayNumber(String value) {
  if (!_isValidDateOnly(value)) return null;
  final parts = value.split('-').map(int.parse).toList();
  return DateTime.utc(parts[0], parts[1], parts[2])
          .difference(DateTime.utc(1970, 1, 1))
          .inDays;
}

bool _isValidDateOnly(String value) {
  final parts = value.split('-');
  if (parts.length != 3 || parts.any((part) => part.isEmpty)) return false;
  final numbers = parts.map(int.tryParse).toList();
  if (numbers.any((number) => number == null)) return false;
  final date = DateTime.utc(numbers[0]!, numbers[1]!, numbers[2]!);
  return date.year == numbers[0] &&
      date.month == numbers[1] &&
      date.day == numbers[2];
}

int? _minutesBetween(DateTime? start, DateTime? end) {
  if (start == null || end == null) return null;
  var difference = end.difference(start).inMinutes;
  if (difference <= 0) difference += 24 * 60;
  return difference;
}

double _clamp(double value, double minimum, double maximum) =>
    value.clamp(minimum, maximum).toDouble();

int _round(num value) => value.round();